docs/routing.md

```diff
@@ -0,0 +1,248 @@
+
+#### Checking the New Destination
+
+Sometimes, a route may rewrite a path and require that its existence is verified in the file system before continuing. On the Vercel [documentation](https://vercel.com/docs/build-output-api/v3/configuration#source-route), it states that the `check` property should trigger entering the `filesystem` and `rewrite` phases.
+
+This can be somewhat problematic.
+
+Requests that should be interpreted as requesting the `RSC` version of a page need to enter the `none` phase to receive a rewrite to the `RSC` path. However, if we assume that `check` should take us back to `filesytem`, instead of the routes that aren't in a phase, we would end up not supporting `RSC` pages properly.
+
+Therefore, when the rewritten path has changed, we send the request back to the `none` phase, so that it can be processed from the start.
+
+In the event that the rewritten path is the same as the current path, we opt to send the request to the next phase when the current phase is not `miss`. This custom check for the `miss` phase is necessary in order to prevent infite loops for requests to `/_next/static/...`, and `/_next/data/...`, as they may be rewritten to themselves when calling `check`. As such, if we are in the `miss` phase, we instead update the status code to 404.
+
+#### Continuing with the Routing Process
+
+After all of the above has been completed, the final part to check for a route is whether it would like us to continue matching. If it does not, we have reached the end of matching for this phase, so we can simply exit out with the `done` response.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkRoute --> CheckSourceMatch{check source match}
+    CheckSourceMatch -->|true| ApplyOverrides[apply overrides]
+    CheckSourceMatch -->|false| EndRoutingSkip[return `skip`]
+    ApplyOverrides --> HasMiddleware{has middleware}
+    HasMiddleware --> |yes| RunMiddlewareResult{run middleware}
+    HasMiddleware --> |no| ApplyHeaders
+    RunMiddlewareResult --> |success| ApplyHeaders
+    RunMiddlewareResult --> |error| EndRoutingError[return `error`]
+    ApplyHeaders --> ApplyStatusCode[apply status code]
+    ApplyStatusCode --> ApplyDestination
+    ApplyDestination --> ShouldCheck{should check filesystem}
+    ShouldCheck --> |yes| PathsEqual{new path is equal}
+    PathsEqual --> |yes + phase = miss| Return404[set status to 404] --> ShouldContinue
+    PathsEqual --> |yes + phase != miss| checkNextPhase[check next phase]
+    PathsEqual --> |false| checkNonePhase[check `none` phase]
+    ShouldCheck --> |false| ShouldContinue{should continue}
+    ShouldContinue --> |true| ReturnNext[return `next`]
+    ShouldContinue --> |false| ReturnDone[return `done`]
+```
```

Comment:
```suggestion
```mermaid
flowchart TD
    checkRoute --> CheckSourceMatch{check source match}
    CheckSourceMatch -->|true| ApplyOverrides[apply overrides]
    CheckSourceMatch -->|false| EndRoutingSkip[`skip`]
    ApplyOverrides --> HasMiddleware{has middleware}
    HasMiddleware --> |yes| RunMiddlewareResult{run middleware}
    HasMiddleware --> |no| ApplyHeadersStatusAndDestination[apply headers\napply status code\napply destination]
    RunMiddlewareResult --> |success| ApplyHeaders[apply headers]
    RunMiddlewareResult --> |error| EndRoutingError[`error`]
    ApplyHeadersStatusAndDestination --> ShouldCheck{should check filesystem}
    ShouldCheck --> |yes| PathsEqual{new path is equal}
    PathsEqual --> |yes\n+\nphase = miss| Return404[set status to 404] --> ShouldContinue
    PathsEqual --> |yes\n+\nphase != miss| checkNextPhase[check next phase]
    PathsEqual --> |false| checkNonePhase[check `none` phase]
    ShouldCheck --> |false| ShouldContinue{should continue}
    ShouldContinue --> |true| ReturnNext[`next`]
    ShouldContinue --> |false| ReturnDone[`done`]
```
```